### PR TITLE
add first version languages images issue #1043

### PR DESCRIPTION
**Pull request by benjamingarcia**
That's ok? [issue #1043](https://github.com/CoderDojo/community-platform/issues/1043)

### CODE REVIEW

**Comment by Wardormeur** [comment]
It also requires to link any image to the languages in conf/prog-lang.js, so we can "blindly" link them together

**Comment by benjamingarcia** [comment]
So, I need to change this list like this : `{'text':'scratch','picture':'/img/languages/scratch.png'}`

**Comment by DanielBrierton** [comment]
Yeah, something along those lines. I'd say probably `image` instead of `picture`, but I'm probably just being picky there.

Ideally, you would update [this template](https://github.com/CoderDojo/cp-zen-platform/blob/master/web/public/templates/profiles/programming-languages-list.dust), and the relevant JS to display the images on the profile.

**Comment by benjamingarcia** [comment]
ok!

**Comment by benjamingarcia** [comment]
I use the [cp-local-development](https://github.com/CoderDojo/cp-local-development), but I don't found right uri to show template.

**Comment by butlerx** [comment]
that template should be part of `localhost:8000/profile` after the children list

**Comment by benjamingarcia** [comment]
all right! I need to create a user. thanks!

**Comment by benjamingarcia** [comment]
Well, what is password for champin1@example.com? it's for test

**Comment by butlerx** [comment]
the test data is all loaded from [here](https://github.com/CoderDojo/cp-users-service/blob/master/test/lib/test-user-data.js)

**Comment by benjamingarcia** [comment]
hye! 
I think It's done. I use now images on profile page.

File: web/public/js/controllers/user-profile-controller.js [hunk 1]
```diff
@@ -173,6 +173,16 @@ function cdUserProfileCtrl($scope, $rootScope, $state, $window, auth, cdUsersSer
     return cdProgrammingLanguagesService.get();
   };
 
+
+  $scope.imageByLanguage = function(language){
+    cdProgrammingLanguagesService.getByLanguage(language,
+      function (responseData) {
+        $scope.profile.programmingLanguagesImageUrl.push(responseData[0]);
+    }, function () {
+      console.error("error when get data")
```
Comment (butlerx) [inline web/public/js/controllers/user-profile-controller.js]: we dont use console.error use logger.error

File: web/public/js/directives/user/cd-profile/view/template.dust [hunk 2]
```diff
@@ -36,7 +36,7 @@
             <projects-list ng-if="!hideChampionProfileBlock('projects') && hasAccess(profile.userTypes, ['champion'])"></projects-list>
             <parents-list ng-if="!hideProfileBlock() && hasAccess(profile.userTypes, ['attendee-u13', 'attendee-o13'])"></parents-list>
             <children-list ng-if="loggedInUserIsCDFAdmin() || hasAccess(profile.userTypes, ['parent-guardian', 'mentor', 'champion'])"></children-list>
-            <programming-languages-list ng-if="!hideProfileBlock('programmingLanguages') && profile.widget.languages"></programming-languages-list>
+            <programming-languages-list></programming-languages-list>
```
Comment (butlerx) [inline web/public/js/directives/user/cd-profile/view/template.dust]: some people may want to not show programming languages its a tick box in the profile

Comment (benjamingarcia) [inline web/public/js/directives/user/cd-profile/view/template.dust]: oh yes. my mistake.

File: web/public/js/controllers/user-profile-controller.js (hunk 1 above)
Comment (Wardormeur) [inline web/public/js/controllers/user-profile-controller.js]: (cian, we don't have logger on the front-end, only the back :p)
But yeah, we avoid having console handling for errors, prefer bootbox to display an alert (we'll come to a generic error handler soon, that's on our github project :))

Comment (benjamingarcia) [inline web/public/js/controllers/user-profile-controller.js]: ok.

Comment (butlerx) [inline web/public/js/controllers/user-profile-controller.js]: wardormeur woops wrong repo :stuck_out_tongue_closed_eyes:

Comment (benjamingarcia) [inline web/public/js/controllers/user-profile-controller.js]: That's I see.. I can use [$log](https://docs.angularjs.org/api/ng/service/$log) if you want?

**Comment by benjamingarcia** [comment]
and now ^^

**Comment by benjamingarcia** [comment]
So, I add $log, and I rollback template.dust.

**Comment by Wardormeur** [comment]
I'm asking myself if requesting multiple time the api for languages is better than requesting it once and just taking what's necessary on the front.
It makes sense when it's a real API, but it's a json file so far :p 
thoughts @DanielBrierton ?

**Comment by benjamingarcia** [comment]
i'm agree with you, that's why I put `language` until service. 
But in my opinion, it's not the same issue. It need to create a real API as you said @Wardormeur.

**Comment by benjamingarcia** [comment]
Hye! I would like to try an update, but since this morning, I have this error : 
`angular.js:13708 TypeError: Cannot read property 'userTypes' of undefined
    at new cdUserProfileCtrl (http://localhost:8000/js/controllers/user-profile-controller.js:96:73)
    at Object.instantiate (http://localhost:8000/components/angular/angular.js:4724:14)
    at $controller (http://localhost:8000/components/angular/angular.js:10249:28)
    at Object.<anonymous> (http://localhost:8000/components/angular-ui-router/release/angular-ui-router.min.js:7:23873)
    at http://localhost:8000/components/angular/angular.js:1240:18
    at invokeLinkFn (http://localhost:8000/components/angular/angular.js:9814:9)
    at nodeLinkFn (http://localhost:8000/components/angular/angular.js:9215:11)
    at compositeLinkFn (http://localhost:8000/components/angular/angular.js:8510:13)
    at publicLinkFn (http://localhost:8000/components/angular/angular.js:8390:30)
    at lazyCompilation (http://localhost:8000/components/angular/angular.js:8728:25) <div ui-view="" class="ng-scope">`

I don't understand. I reset my workspace and I have this error on my fork as in official master

**Comment by Wardormeur** [comment]
Could you tell us when does it happens? I haven't been able to reproduce it on the profile page

**Comment by benjamingarcia** [comment]
1. reset all (except databases)  `rm -fr workspace-zen`, and `./localdev.js init zen`
2. restart env (`./localdev.js run zen`)
3. Logon on app
4. Go to profile

**Comment by benjamingarcia** [comment]
I think, my code is ready, but I can't test 😞

**Comment by Wardormeur** [comment]
We'll probably come back to you next week to find out what's happening, we're quite booked right now :p

**Comment by benjamingarcia** [comment]
I already do this. 
I reset DB too and try again.

**Comment by benjamingarcia** [comment]
Allright! I finish! 🎉

**Comment by Wardormeur** [comment]
Nearly there.
First, move the programmingLanguageUrl out of the profile, elsewhat it'll crash on save of the profile.
Then, I'd consider reorganisaing programmingLanguageUrl on a key:url base in order to be able to have a name for accessibility on those images :)
Finally, profile.widget.programmingLanguages seems to not exist anymore :p

**Comment by benjamingarcia** [comment]
allright 👍

**Comment by benjamingarcia** [comment]
I don't do the map. How do you what it? Like this : `{'scratch':'url'}`

**Comment by benjamingarcia** [comment]
Hye! Now I return an object like this : `{imageUrl:language.image, text:language.text}`. 
I don't understand why I can't remove map function. 
Without it this code can be better because this object is the same as `programmingLanguages.js`. But I don't have much time to investigate before 31th october

File: web/public/js/directives/user/cd-profile/view/template.dust [hunk 3]
```diff
@@ -36,7 +36,7 @@
             <projects-list ng-if="!hideChampionProfileBlock('projects') && hasAccess(profile.userTypes, ['champion'])"></projects-list>
             <parents-list ng-if="!hideProfileBlock() && hasAccess(profile.userTypes, ['attendee-u13', 'attendee-o13'])"></parents-list>
             <children-list ng-if="loggedInUserIsCDFAdmin() || hasAccess(profile.userTypes, ['parent-guardian', 'mentor', 'champion'])"></children-list>
-            <programming-languages-list ng-if="!hideProfileBlock('programmingLanguages') && profile.widget.languages"></programming-languages-list>
+              <programming-languages-list ng-if="!hideProfileBlock('programmingLanguages') && profile.widget.languages"></programming-languages-list>
```
Comment (Wardormeur) [inline web/public/js/directives/user/cd-profile/view/template.dust]: Does it show up in your profile? Last time i checked, profile.widget.language was always undefined

**Comment by Wardormeur** [comment]
Hey, I'll try to have a look at it tomorrow :)

File: web/public/js/directives/user/cd-profile/view/template.dust (hunk 3 above)
Comment (Wardormeur) [inline web/public/js/directives/user/cd-profile/view/template.dust]: Well, it's better to simply have a profile with badges, switch the parameter to hide the block to check the visibility, and one without badges at all :p

**Comment by Wardormeur** [comment]
Thanks for the linting :+1:

**Comment by benjamingarcia** [comment]
Hello. so? it's 31th October. It's last day for hacktoberfest and I don't have much time. 
Please 🙏  close this issue. I can continue to work on it with pleasure. But close issue #1043 ;-)

**Comment by Wardormeur** [comment]
Don't worry; Hacktoberfest takes PR into account, even if not merged :)
I'll have a look again during the week, i promise :)

**Comment by benjamingarcia** [comment]
all Right!

File: web/public/templates/profiles/programming-languages-list.dust [hunk 4]
```diff
@@ -1,8 +1,6 @@
 <div class="profile-section cd-programming-languages">
   <h3 class="cd-section-header"><a id="#prog"></a><a href="#prog">{@i18n key="Programming Languages"/}</a></h3>
-  <div>
-    <ul ng-repeat="language in profile.widget.programmingLanguages" class="list-unstyled">
-      <li class="col-md-6"> <span class="badge">{{ language }}</span></li>
-    </ul>
+  <div class="col-sm-2" ng-repeat="language in programmingLanguagesImageUrl">
```
Comment (Wardormeur) [inline web/public/templates/profiles/programming-languages-list.dust]: please define a size for xs display, i think keeping the same ratio over a small display is fine, a prog language isn't the primary info of a profile

File: web/public/js/controllers/user-profile-controller.js [hunk 5]
```diff
@@ -273,79 +292,43 @@ function cdUserProfileCtrl($scope, $rootScope, $state, $window, auth, cdUsersSer
...
+            childrenCopy.gender = child.gender;
+            childrenCopy.email = child.email;
+            childrenCopy.parents = [profileCopy.id];
+            if (getAge(child.dateOfBirth) >= 13) {
+              childrenCopy.userTypes = ['attendee-o13'];
             } else {
               $scope.profile = profile;
               $scope.profile.private =  $scope.profile.private ? "true" : "false";
-              alertService.showAlert(message);
+              //alertService.showAlert(message);
```
Comment (Wardormeur) [inline web/public/js/controllers/user-profile-controller.js]: why removed?

File: web/public/js/directives/user/cd-profile/view/template.dust (hunk 3 above)
Comment (Wardormeur) [inline web/public/js/directives/user/cd-profile/view/template.dust]: confirmed, you need to replace profile.widget.languages by programmingLanguagesImageUrl > 0

File: web/public/js/controllers/user-profile-controller.js [hunk 6]
```diff
@@ -273,79 +292,43 @@ function cdUserProfileCtrl($scope, $rootScope, $state, $window, auth, cdUsersSer
...
+          localStorage.setItem('children', 'true');
+          async.mapSeries($scope.profile.children, function (child, doneChild) {
+            childrenCopy.name = child.name;
+            childrenCopy.alias = child.alias;
+            childrenCopy.dob = child.dateOfBirth;
+            childrenCopy.gender = child.gender;
+            childrenCopy.email = child.email;
+            childrenCopy.parents = [profileCopy.id];
+            if (getAge(child.dateOfBirth) >= 13) {
+              childrenCopy.userTypes = ['attendee-o13'];
```
Comment (Wardormeur) [inline web/public/js/controllers/user-profile-controller.js]: Not sure what happened here (and the ~50 next lines), but you mixed up the final callback with one of the series actions. You also removed the conditional assignation to u13 userType.
Why? :p

File: web/config/programmingLanguages.js [hunk 7]
```diff
@@ -1,29 +1,107 @@
...
+    'text': 'Lua',
+    'image': '/img/languages/lua.png'
+  },
+  {
+    'text': 'Bash',
+    'image': '/img/languages/bash.png'
+  },
+  {
+    'text': 'Haskell',
+    'image': '/img/languages/haskell.png'
```
Comment (Wardormeur) [inline web/config/programmingLanguages.js]: capital H

File: web/public/templates/profiles/programming-languages-list.dust (hunk 4 above)
Comment (Wardormeur) [inline web/public/templates/profiles/programming-languages-list.dust]: Also, it requires to have a row to contains all the cols, elsewhat next elements will float on the right side (try having a dojo and only 1 prog lang)

File: web/public/js/controllers/user-profile-controller.js (hunk 6 above)
Comment (benjamingarcia) [inline web/public/js/controllers/user-profile-controller.js]: well, in fact, it's not me, it's during the rebase. 😄

File: web/config/programmingLanguages.js (hunk 7 above)
Comment (benjamingarcia) [inline web/config/programmingLanguages.js]: ok

File: web/public/js/controllers/user-profile-controller.js (hunk 5 above)
Comment (benjamingarcia) [inline web/public/js/controllers/user-profile-controller.js]: same as before, it was during the rebase. I don't know why. 😕

Comment (benjamingarcia) [inline web/public/js/controllers/user-profile-controller.js]: because ```message``` is unknown it was the *jshint* error last week.

**Comment by Wardormeur** [comment]
Updated by @grabartley and  @campbs.
Thanks all of you :)

**Pull request closed without merging**